src/contexts/WalletContext.tsx

```diff
@@ -0,0 +1,208 @@
+import type { ReactNode } from 'react';
+import React, {
+  createContext,
+  useContext,
+  useState,
+  useCallback,
+  useMemo,
+  useEffect,
+} from 'react';
+import type { Chain } from '@wormhole-foundation/sdk';
+import { chainToPlatform } from '@wormhole-foundation/sdk';
+import { useDispatch } from 'react-redux';
+import config from 'config';
+import type { Wallet, WormholeConnectWalletProvider } from 'utils/wallet';
+import { TransferWallet } from 'utils/wallet';
+import {
+  clearWallet,
+  connectWallet as connectSourceWallet,
+  connectReceivingWallet,
+  swapWallets as swapWalletsAction,
+} from 'store/wallet';
+import { setAmount, swapInputs } from 'store/transferInput';
+import { InternalWalletProvider } from 'utils/wallet/provider';
+
+interface WalletContextType {
+  connectWallet: (
+    chain: Chain,
+    type: TransferWallet,
+    connectLastUsed?: boolean,
+  ) => Promise<Wallet>;
+  swapWallets: () => boolean;
+  disconnectWallet: (chain: Chain, type: TransferWallet) => Promise<void>;
+  isConnecting: boolean;
+  walletProvider: WormholeConnectWalletProvider;
+  isInternalProvider: boolean;
+}
+
+export const WalletContext = createContext<WalletContextType | undefined>(
+  undefined,
+);
+
+interface WalletProviderProps {
+  children: ReactNode;
+  provider: WormholeConnectWalletProvider;
+  isInternal: boolean;
+}
+
+export const WalletProvider: React.FC<WalletProviderProps> = ({
+  children,
+  provider,
+  isInternal,
+}) => {
+  const [isConnecting, setIsConnecting] = useState(false);
+  const dispatch = useDispatch();
+
+  const setupWalletConnection = useCallback(
+    (wallet: Wallet, chain: Chain, type: TransferWallet): void => {
+      console.log(`Setting up wallet connection for ${chain} (${type})...`);
+
+      const address = wallet.getAddress();
+      if (!address) {
+        throw new Error('Wallet is not connected');
+      }
+
+      config.triggerEvent({
+        type: 'wallet.connect',
+        details: {
+          side: type,
+          chain: chain,
+          wallet: wallet.getName().toLowerCase(),
+        },
+      });
+
+      const payload = {
+        address,
+        type: chainToPlatform(chain),
+        icon: wallet.getIcon(),
+        name: wallet.getName(),
+      };
+
+      if (type === TransferWallet.SENDING) {
+        dispatch(connectSourceWallet(payload));
```

Comment:
Ideally all wallet state would exist in this context. But I don't want to refactor too much in this PR